c++/src/kj/compat/http.c++

```diff
@@ -1547,6 +1547,466 @@ private:
+
+  kj::Own<kj::AsyncIoStream> stream;
+  kj::Maybe<WebSocket::MaskKeyGenerator&> maskKeyGenerator;
+
+  bool sendClosed = false;
+  bool currentlySending = false;
+  Header sendHeader;
+  kj::ArrayPtr<const byte> sendParts[2];
+
+  kj::Maybe<kj::Array<byte>> queuedPong;
+  // If a Ping is received while currentlySending is true, then queuedPong is set to the body of
+  // a pong message that should be sent once the current send is complete.
+
+  kj::Maybe<kj::Promise<void>> sendingPong;
+  // If a Pong is being sent asynchronously in response to a Ping, this is a promise for the
+  // completion of that send.
+
+  uint fragmentOpcode = 0;
+  kj::Vector<kj::Array<byte>> fragments;
+  // If `fragments` is non-empty, we've already received some fragments of a message.
+  // `fragmentOpcode` is the original opcode.
+
+  uint recvAvail = 0;
+  kj::Array<byte> recvBuffer;
+
+  kj::Promise<void> sendImpl(byte opcode, kj::ArrayPtr<const byte> message) {
+    KJ_REQUIRE(!sendClosed, "WebSocket already closed");
+    KJ_REQUIRE(!currentlySending, "another message send is already in progress");
+
+    currentlySending = true;
+
+    KJ_IF_MAYBE(p, sendingPong) {
+      // We recently sent a pong, make sure it's finished before proceeding.
+      auto promise = p->then([this, opcode, message]() {
+        currentlySending = false;
+        return sendImpl(opcode, message);
+      });
+      sendingPong = nullptr;
+      return promise;
+    }
+
+    sendClosed = opcode == OPCODE_CLOSE;
+
+    Mask mask(maskKeyGenerator);
+
+    kj::Array<byte> ownMessage;
+    if (!mask.isZero()) {
+      // Sadness, we have to make a copy to apply the mask.
+      ownMessage = kj::heapArray(message);
+      mask.apply(ownMessage);
+      message = ownMessage;
+    }
+
+    sendParts[0] = sendHeader.compose(true, opcode, message.size(), mask);
+    sendParts[1] = message;
+
+    auto promise = stream->write(sendParts);
+    if (!mask.isZero()) {
+      promise = promise.attach(kj::mv(ownMessage));
+    }
+    return promise.then([this]() {
+      currentlySending = false;
+
+      // Send queued pong if needed.
+      KJ_IF_MAYBE(q, queuedPong) {
+        kj::Array<byte> payload = kj::mv(*q);
+        queuedPong = nullptr;
+        queuePong(kj::mv(payload));
+      }
+    });
+  }
+
+  void queuePong(kj::Array<byte> payload) {
+    if (currentlySending) {
+      // There is a message-send in progress, so we cannot write to the stream now.
+      //
+      // Note: According to spec, if the server receives a second ping before responding to the
+      //   previous one, it can opt to respond only to the last ping. So we don't have to check if
+      //   queuedPong is already non-null.
+      queuedPong = kj::mv(payload);
+    } else KJ_IF_MAYBE(promise, sendingPong) {
+      // We're still sending a previous pong. Wait for it to finish before sending ours.
+      sendingPong = promise->then(kj::mvCapture(payload, [this](kj::Array<byte> payload) mutable {
```

Comment:
True, though that will change as soon as we switch to C++14 and can change the kj::mvCapture() into a native capture.

I wish there were a way that we could declare all promise continuations mutable, because const is never what you want for them, and it results in confusing error messages. So, often I write `mutable` indiscriminately to avoid confusion later.